## Project Decomposition



### Preprocess the data:

1. Identify and fill in missing values
2. Replace the real number data type with the integer type
3. Delete duplicate data
4. Categorize the data

### Data Analysis

1. Is there a connection between having kids and repaying a loan on time?
2. Is there a connection between marital status and repaying a loan on time?
3. Is there a connection between income level and repaying a loan on time?
4. How do different loan purposes affect on-time loan repayment?
5. Interpret your answers. Explain what the results you obtained mean.


## Table of Contents

* [Chapter 1: Iporting Data File & Preprocessing](#chapter1)
    * [Section 1.1: Description of the data](#section_1_1)
    * [Section 1.2: Data Format and Info](#section_1_2)
      
 
* [Chapter 2: Handling Missing Values](#chapter2)
    * [Section 2.1: Missing Values Identification](#section_2_1)
    * [Section 2.2: Filling Missing Values](#section_2_2)
    * [Section 2.3: Approach to Fill Missing Values](#section_2_3)
    

* [Chapter 3: Handling Duplicates](#chapter3)
    * [Section 3.1: Changing Data Type](#section_3_1)
    * [Section 3.2: Deleting Duplicates ](#section_3_2)
    * [Section 3.3: Reasoning ](#section_3_3)

    
* [Chapter 4: Categorizing the Data](#chapter4)
    * [Section 4.1: Categorizing the Income Level](#section_4_1)
    * [Section 4.2: Categorizing the Purpose of Loan ](#section_4_2)
    * [Section 4.3: Reasoning ](#section_4_3)
    
    
* [Chapter 5: Let's Find Answer](#chapter5)
    * [Section 5.1: Is there a connection between having kids and repaying a loan on time?](#section_5_1)
    * [Section 5.2: Is there a connection between marital status and repaying a loan on time?](#section_5_2)
    * [Section 5.3: Is there a connection between income level and repaying a loan on time?](#section_5_3)
    * [Section 5.4: How do different loan purposes affect on-time loan repayment?](#section_5_4)


* [Chapter 6: Conclusion](#chapter6)

### Chapter 1: Iporting Data File & Preprocessing <a class="anchor" id="chapter1"></a>

In [1]:
import pandas as pd
import numpy as np

In [2]:

try:
    data=pd.read_csv('credit_scoring_eng.csv')
except:
    data=pd.read_csv(r"C:\Users\engrz\Python Projects\practicum\credit_scoring_eng.csv")
data.head(10)


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

#### Section 1.1: Description of the data <a class="anchor" id="section_1_1"></a>

1. __children:__ the number of children in the family
2. __days_employed:__ how long the customer has been working
3. __dob_years:__ the customer’s age
3. __education:__ the customer’s education level
4. __education_id:__ identifier for the customer’s education
5. __family_status:__ the customer’s marital status
6. __family_status_id:__ identifier for the customer’s marital status
7. __gender:__ the customer’s gender
8. __income_type:__ the customer’s income type
9. __debt:__ whether the customer has ever defaulted on a loan
10. __total_income:__ monthly income
11. __purpose:__ reason for taking out a loan

#### Section 1.2: Data Format and Info <a class="anchor" id="section_1_2"></a>

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


This dataset contains 21525 rows and 12 columns. There are two columns (days_employed, total_income) with missing values. Most of the datatype is according to requirement of the analysis but days_employed and total_income datatype will be change from float to interger to complete the requirements.

In [4]:
data.head(5)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

### Chapter 2: Handling Missing Values <a class="anchor" id="chapter2"></a>

#### Section 2.1: Missing Values Identification <a class="anchor" id="section_2_1"></a>

In [5]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

In [6]:
data.describe(include=np.object)

<ipython-input-6-289c2ff97710>:1: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. 
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  data.describe(include=np.object)


education family_status gender income_type           purpose
count                 21525         21525  21525       21525             21525
unique                   15             5      3           8                38
top     secondary education       married      F    employee  wedding ceremony
freq                  13750         12380  14236       11119               797

In [7]:
print(data.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


The describe() function has shown that the 2 columns (days_employed,total_income) have 2174 missing entries.

In [8]:
print(data.isna().sum()/len(data)*100) # this is use to calculate the ratio of the missing values and it is about 10% of total data.

children             0.000000
days_employed       10.099884
dob_years            0.000000
education            0.000000
education_id         0.000000
family_status        0.000000
family_status_id     0.000000
gender               0.000000
income_type          0.000000
debt                 0.000000
total_income        10.099884
purpose              0.000000
dtype: float64


In [9]:
data['days_employed']=data['days_employed'].abs() # this function use to change negative values to absolute.

#### Section 2.2: Filling Missing Values <a class="anchor" id="section_2_2"></a>

To fill the misssing values the median were calculated of the column and using fillna() method na values were replaced by the median of the column. The median was perferred rather than mean because there would be skewed distribution.

In [10]:
days_employed_med=data['days_employed'].median()
total_income_med=data['total_income'].median()
print(days_employed_med)
print(total_income_med)

2194.220566878695
23202.87


In [11]:
#check the total no. of values after filling using count()
data['days_employed'].fillna(value=days_employed_med, inplace=True)
data['total_income'].fillna(value=total_income_med, inplace=True)
print(data['days_employed'].count())
print(data['total_income'].count())

21525
21525


### Chapter 3: Handling Duplicates <a class="anchor" id="chapter3"></a>

__Removal of dataset's skewed values__

In the column children of dataset there is typo and it seams that 1 and 20 were filled by -1 and 20 mistakenly. We will take absolute of the values to and replace with the actual corresponding value.  

In [12]:
print(data.children.unique())

[ 1  0  3  2 -1  4 20  5]


The -ve value in children column is replaced by the an abs() value.

In [13]:
data['children']=data['children'].abs()

In [14]:
print(data.children.unique())

[ 1  0  3  2  4 20  5]


In [15]:
data.loc[data['children'] == 20,'children']= 2

In [16]:
print(data.children.unique())

[1 0 3 2 4 5]


In [17]:
data.head(2)

children  days_employed  dob_years            education  education_id  \
0         1    8437.673028         42    bachelor's degree             0   
1         1    4024.803754         36  secondary education             1   

  family_status  family_status_id gender income_type  debt  total_income  \
0       married                 0      F    employee     0     40620.102   
1       married                 0      F    employee     0     17932.802   

                 purpose  
0  purchase of the house  
1           car purchase

After fixing children and days_employed columns of dataset, the very next column is date of birth dataset or age of the customer dataset. When the value_counts method was applied to the column it was observed that the age was starting from the 21 but there were also some values with age zero. The zero entry age has 101 values. The possible reason which was detected the this age group has mostly college or school level education. The reason for missing value whould be that relationship in very early age but this observation can be rejected on the basis when the customer failed to repay the loan.

In [18]:
print(data['dob_years'].value_counts())

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64


In [19]:
data[data['dob_years']==0]
missing_age=data[data.dob_years==0]
missing_age.head()
len(missing_age)

101

The relationship of age=0 values were observed with other columns. The mostly people has bachelor or secondary school education. One observation is that the these people are not really belong to one category but randomly distributed in different group of people.So, the possible reason could be that people were not willing to share date of birth or missid by the data entry person.

In [20]:
len(missing_age)

101

In [21]:
len(missing_age)/len(data['dob_years'])*100

0.4692218350754936

In [22]:
# checking relation between income type. 
missing_age['income_type'].value_counts()

employee         55
retiree          20
business         20
civil servant     6
Name: income_type, dtype: int64

In [23]:
missing_age.groupby('education')['dob_years'].sum() # during this step, it was noticed that there are same education level values 
                                                    # with lower and upper string. 

education
BACHELOR'S DEGREE      0
Bachelor's Degree      0
SECONDARY EDUCATION    0
Secondary Education    0
bachelor's degree      0
secondary education    0
some college           0
Name: dob_years, dtype: int64

In [24]:
missing_age.groupby('family_status')['dob_years'].sum()

family_status
civil partnership    0
divorced             0
married              0
unmarried            0
widow / widower      0
Name: dob_years, dtype: int64

In [25]:
print(missing_age.groupby('children')['dob_years'].sum())

children
0    0
1    0
2    0
3    0
Name: dob_years, dtype: int64


In [26]:
retire_Avg=data.loc[data['income_type'] == 'retiree','dob_years'].median()
retire_Avg

60.0

In [27]:
employee_Avg=data.loc[data['income_type'] == 'employee','dob_years'].median()
#print(employee_Avg)
business_Avg=data.loc[data['income_type'] == 'business','dob_years'].median()
#print(business_Avg)
civil_Avg=data.loc[data['income_type'] == 'civil servant','dob_years'].median()
#print(civil_Avg)
employed= [employee_Avg,business_Avg,civil_Avg]
Avg_employed=np.mean(employed).astype(np.int64)
Avg_employed

39

In [28]:
data[(data['dob_years'] == 0) & (data['income_type'] == 'retiree')].replace(0,retire_Avg)#, inplace=True
len(data[(data['dob_years'] == 0) & (data['income_type'] == 'retiree')].replace(0,retire_Avg))

20

#### Three approaches to deal with missing age values.
1. The percentage of missing values is very less compare to the dataset. So we can drop these values as it will not relly impact the final result.

2. Devide the missing ages rows in two groups of Retiree and Employees with groupby income type. Find the median of these two groups and fill respected values.

3. Fill missing values with the median() of the column dob_years.

The last approach was choosed

The education column data string are entered both in lower and upper string format we will apply str.lower() method to deal with this error.

In [29]:
print(data['education'].unique())

["bachelor's degree" 'secondary education' 'Secondary Education'
 'SECONDARY EDUCATION' "BACHELOR'S DEGREE" 'some college'
 'primary education' "Bachelor's Degree" 'SOME COLLEGE' 'Some College'
 'PRIMARY EDUCATION' 'Primary Education' 'Graduate Degree'
 'GRADUATE DEGREE' 'graduate degree']


In [30]:
data['education'] = data['education'].str.lower()
print(data['education'].unique())

["bachelor's degree" 'secondary education' 'some college'
 'primary education' 'graduate degree']


In [31]:
data['education'].count()

21525

In [32]:
data['family_status'].unique() #check any duplicate string in family_status.

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

In [33]:
data['income_type'].unique() # check any duplicate string in income_type

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

In [34]:
#print(data.groupby('family_status')['family_status_id'].sum())

In [35]:
data.isnull().sum() #check any null value reaming in dataset. there is no missing value so dataset is ready for analysis

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [36]:
print(data.isnull().values.any())

False


In [37]:
 
data['children'].value_counts() # most people has 0 children. or bank gives loan mostly to no children parents?

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

In [38]:
from nltk.stem import SnowballStemmer 
english_stemmer = SnowballStemmer('english')



In [39]:
data['debt'].value_counts() ####################

0    19784
1     1741
Name: debt, dtype: int64

#### Section 3.1: Changing Data Type <a class="anchor" id="section_3_1"></a>

In [40]:
### Replace the real number data type with the integer type
data['days_employed']=data['days_employed'].astype('int64')
data['total_income']=data['total_income'].astype('int64')

In [41]:
data['days_employed']=data['days_employed'].astype('int64')
data['dob_years']=data['dob_years'].astype('int64')
#data['days_employed'].apply(type).unique()
data['dob_years'].apply(type).unique()

array([<class 'int'>], dtype=object)

The dataset dtype was changed from float to integer so that days_employed could be a even number rather than a fraction.

In [42]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


#### Section 3.2: Deleting Duplicates <a class="anchor" id="section_3_2"></a>

In [43]:
len(data[data.duplicated(keep=False)])


137

In [44]:
duplicate=data[data.duplicated()]
duplicate.head(5)

children  days_employed  dob_years            education  education_id  \
2849         0           2194         41  secondary education             1   
3290         0           2194         58  secondary education             1   
4182         1           2194         34    bachelor's degree             0   
4851         0           2194         60  secondary education             1   
5557         0           2194         58  secondary education             1   

          family_status  family_status_id gender income_type  debt  \
2849            married                 0      F    employee     0   
3290  civil partnership                 1      F     retiree     0   
4182  civil partnership                 1      F    employee     0   
4851  civil partnership                 1      F     retiree     0   
5557  civil partnership                 1      F     retiree     0   

      total_income                              purpose  
2849         23202  purchase of the house for my family  
3290         23202                    to have a wedding  
4182         23202                     wedding ceremony  
4851         23202                     wedding ceremony  
5557         23202                    to have a wedding

When you're checking for duplicates, duplicated() method returns the duplicate itself.
https://queirozf.com/entries/pandas-dataframe-examples-duplicated-data.


#### Section 3.3: Reasoning <a class="anchor" id="section_3_3"></a>


There is a relation between duplicate values and the days_employed & total_income missing valaues. We observed this after filling valaues of days_employed & total_income that all the values we filled are now duplicate rows in dataset. 

The reason would be when the data was  collected the age was missing for some customers and again data has collected for these customers but the previous values were not updated or deleted after re enentring the missing values. 

In [45]:
data=data.drop_duplicates()

In [46]:
data = data.reset_index(drop=True) # reseting the index after the dropping duplicate values.
data.info()                       # data is ready for next steps :)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21454 non-null  int64 
 1   days_employed     21454 non-null  int64 
 2   dob_years         21454 non-null  int64 
 3   education         21454 non-null  object
 4   education_id      21454 non-null  int64 
 5   family_status     21454 non-null  object
 6   family_status_id  21454 non-null  int64 
 7   gender            21454 non-null  object
 8   income_type       21454 non-null  object
 9   debt              21454 non-null  int64 
 10  total_income      21454 non-null  int64 
 11  purpose           21454 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


### Chapter 4: Categorizing the Data <a class="anchor" id="chapter4"></a>

In [47]:
from nltk.stem import SnowballStemmer 
english_stemmer = SnowballStemmer('english') 

#### Section 4.1: Categorizing the Income Level  <a class="anchor" id="section_4_1"></a>

The total_income was categorize in four levels.
1. Low income <=10,000
2. Low middle income <=20,000
3. Middle income <=30,000
4. High Income >30,000


In [48]:
#df['total_income'] = pd.to_numeric(df['total_income'], errors='coerce')
 
data.groupby(['income_type'])['total_income'].median()
#data['total_income'] = data['total_income'].astype('float')
 
def income_cat(total_income):
    """
    The function returns the income category according to the total_income value, using the following rules:
Low-income - under 10,000
Lower-middle income - 10,001 - 20,000
Upper-middle income - 20,001 - 30,000
High income > 30,001
    """
 
    if total_income <= 10000:
        return 'low_income'
    if 10001 <= total_income   <= 20000:
        return 'lower_mid_income'
    if 20001 <= total_income <= 30000:
        return 'middle_income'
    return 'high_income'
 
data['income_level'] = data['total_income'].apply(income_cat)
 
 
data.head()
 

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   

                   purpose      income_level  
0    purchase of the house       high_income  
1             car purchase  lower_mid_income  
2    purchase of the house     middle_income  
3  supplementary education       high_income  
4        to have a wedding     middle_income

#### Section 4.2: Categorizing the Purpose of Loan<a class="anchor" id="section_4_2"></a>

__The general overview.__


The first observation is that the purpose column look like a 'feedback/comment' option in form filling during loan process. 
There are different string values for the same purpose of laon, i.e. to have a wedding & having a wedding etc. There are many values i.e. car, car purpose, and cars.

In [49]:
data.groupby('purpose')['debt'].sum()

purpose
building a property                         54
building a real estate                      48
buy commercial real estate                  47
buy real estate                             43
buy residential real estate                 41
buying a second-hand car                    36
buying my own car                           46
buying property for renting out             52
car                                         42
car purchase                                42
cars                                        44
construction of own property                42
education                                   32
getting an education                        37
getting higher education                    46
going to university                         43
having a wedding                            64
housing                                     46
housing renovation                          35
housing transactions                        48
profile education                           44
prope

In [50]:
data['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

In [51]:
len(data['purpose'].unique())

38

In [52]:
print(data['purpose'].value_counts())

wedding ceremony                            791
having a wedding                            768
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
housing                                     646
purchase of the house                       646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
buy real estate                             621
purchase of my own house                    620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

The loan purpose dataset is now categorize in five main groups. All the entries related to wedding are put in wedding, laon for house is put in house, for property bases in property, for eduaction bases in education and all related to real states are combined in real state.

In [53]:
data['purpose'] = data['purpose'].replace(['second-hand car purchase'],'buying a second-hand car')
data['purpose'] = data['purpose'].replace(['car purchase'],'purchase of a car')

In [54]:
import numpy as np
from nltk.stem import SnowballStemmer 
english_stemmer = SnowballStemmer('english') 
queries = data['purpose']
#data['category'] = np.nan
queries

for i, query in enumerate(data['purpose']):
    #print(i,query)
    for word in query.split(" "):
        #print(word)
        stemmed_word = english_stemmer.stem(word)
        if (stemmed_word == 'wed') | (stemmed_word == 'ceremoni'):
            data.loc[i,'category'] = 'Wedding'
        #else:
            #print('ok')
        elif stemmed_word == 'properti':
            data.loc[i,'category'] = 'Property' 
        elif (stemmed_word == 'univers') | (stemmed_word == 'educ'):
            data.loc[i,'category'] = 'Education'
        elif (stemmed_word == 'car'):
            data.loc[i,'category'] = 'Buy a Car'
        else:
            data.loc[i,'category'] = 'Property'
'''for i,x in enumerate(data):
    print(x)
    data.dropna(subset = [x], inplace=True)'''
data.head()

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   

                   purpose      income_level   category  
0    purchase of the house       high_income   Property  
1        purchase of a car  lower_mid_income  Buy a Car  
2    purchase of the house     middle_income   Property  
3  supplementary education       high_income  Education  
4        to have a wedding     middle_income    Wedding

In [55]:
data.tail()

children  days_employed  dob_years            education  education_id  \
21449         1           4529         43  secondary education             1   
21450         0         343937         67  secondary education             1   
21451         1           2113         38  secondary education             1   
21452         3           3112         38  secondary education             1   
21453         2           1984         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
21449  civil partnership                 1      F    business     0   
21450            married                 0      F     retiree     0   
21451  civil partnership                 1      M    employee     1   
21452            married                 0      M    employee     1   
21453            married                 0      F    employee     0   

       total_income               purpose      income_level   category  
21449         35966  housing transactions       high_income   Property  
21450         24959     purchase of a car     middle_income  Buy a Car  
21451         14347              property  lower_mid_income   Property  
21452         39054     buying my own car       high_income  Buy a Car  
21453         13127          to buy a car  lower_mid_income  Buy a Car

#### Section 4.3: Reasoning <a class="anchor" id="section_4_3"></a>

In categorizing the loan purpose column, it was observed that mostly the loans were fall into four main categories which are buying a car, property, wedding and education. The column indices i.e. second-hand car purchase is similar to buying a second-hand car and car purchase is same as purchase a car. These two indices were combine using replace name method. This was easy to stemm purposes. Afterward the purposes were combined in category columns applying the stemming method as in above function.The values were traced by head and tail function to identify any uncategorize indicies.

### Chapter 5: Credit Report <a class="anchor" id="chapter5"></a>

#### Section 5.1: Is there a connection between having kids and repaying a loan on time?<a class="anchor" id="section_5_1"></a>

__Approach:__
    To get the answer, first we will devide the children dataset in family group. So the unique() values for children dataset are 0,1,2,3,4 and 5. 
    
We will devide families in two groups. 
    
1. Small Family: No. of Kids 0-2
    
2. large Family: No. of Kids 3-5

After that we will observe which family group has high default rate. 

__Default Rate Formula__ = No.of Defaults / Total No. of Loans

In [56]:
kid_debt = data[['children','debt']]


def children_debts(kid_debt):
    kids = kid_debt['children']
    debts_status = kid_debt['debt']
 
    if kids <= 2:
        if debts_status == 0:
            return 'small family and happy'
        return 'small family with debts'
    if kids >= 3:
        if debts_status == 0:
            return 'large family and happy'
        return 'large family with debts'


data['debt_children_category']=data.apply(children_debts,axis=1)


In [57]:
data.head()

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   

                   purpose      income_level   category  \
0    purchase of the house       high_income   Property   
1        purchase of a car  lower_mid_income  Buy a Car   
2    purchase of the house     middle_income   Property   
3  supplementary education       high_income  Education   
4        to have a wedding     middle_income    Wedding   

   debt_children_category  
0  small family and happy  
1  small family and happy  
2  small family and happy  
3  large family and happy  
4  small family and happy

In [58]:
data['debt_children_category'].value_counts()

small family and happy     19364
small family with debts     1710
large family and happy       349
large family with debts       31
Name: debt_children_category, dtype: int64

In [59]:
data['debt_children_category']
small_family_default=len(data[data.debt_children_category == 'small family with debts'])/(len(data[data.debt_children_category == 'small family with debts'])+(len(data[data.debt_children_category == 'small family and happy'])))*100
small_family_default

8.114264022017652

In [60]:
data['debt_children_category']
large_family_default=len(data[data.debt_children_category == 'large family with debts'])/(len(data[data.debt_children_category == 'large family with debts'])+(len(data[data.debt_children_category == 'large family and happy'])))*100
large_family_default

8.157894736842106

# Answer:

In [61]:
children_default={'Family Level':['Small Family','Large Family'], 'Default Rate':[int(small_family_default),int(large_family_default)]}
children_default=pd.DataFrame(data=children_default)
children_default

Family Level  Default Rate
0  Small Family             8
1  Large Family             8

That'S seem the relation between default rate of both family groups is same.

#### Section 5.2: Is there a connection between marital status and repaying a loan on time?<a class="anchor" id="section_5_2"></a>

__Approach:__

For the purpose data main dataset column was not divide into any further category. First the dataset columns are extracted into marital_status dataFrame.

In [62]:
#married=0, civil_partnership=1, widow=2,  divorced=3, unmarried=4

marital_status = data[['family_status_id','debt']]
print(marital_status.head())


   family_status_id  debt
0                 0     0
1                 0     0
2                 0     0
3                 0     0
4                 1     0


In [63]:
''' The function below explain that if the customer is married and there is not debt than he/she is married and happy
or vise versa. this approach is same for every marital status person'''

def marital_status_debt(marital_status):
    status = marital_status['family_status_id']
    no_debts = marital_status['debt']
 
    if status == 0:
        if no_debts == 0:
            return 'married and happy'
        return 'married and default'
    if status == 1:
        if no_debts == 0:
            return 'partner and happy'
        return 'partner and default'
    if status == 2:
        if no_debts == 0:
            return 'widow and happy'
        return 'widow but default'
    if status == 3:
        if no_debts == 0:
            return 'divorced and happy'
        return 'divorced but default'
    if status == 4:
        if no_debts == 0:
            return 'single and happy'
        return 'single but default'


data['marital_status_debt']=data.apply(marital_status_debt,axis=1)
data.head()

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   

                   purpose      income_level   category  \
0    purchase of the house       high_income   Property   
1        purchase of a car  lower_mid_income  Buy a Car   
2    purchase of the house     middle_income   Property   
3  supplementary education       high_income  Education   
4        to have a wedding     middle_income    Wedding   

   debt_children_category marital_status_debt  
0  small family and happy   married and happy  
1  small family and happy   married and happy  
2  small family and happy   married and happy  
3  large family and happy   married and happy  
4  small family and happy   partner and happy

In [64]:
data['marital_status_debt'].value_counts()

married and happy       11408
partner and happy        3763
single and happy         2536
divorced and happy       1110
married and default       931
widow and happy           896
partner and default       388
single but default        274
divorced but default       85
widow but default          63
Name: marital_status_debt, dtype: int64

In [65]:
marital_status = data[['family_status_id','debt']]


marital_debt=marital_status.groupby('family_status_id').mean().sort_values('debt',ascending=False)*100
marital_debt

debt
family_status_id          
4                 9.750890
1                 9.347145
0                 7.545182
3                 7.112971
2                 6.569343

In [66]:
data['marital_status_debt'].unique()

array(['married and happy', 'partner and happy', 'partner and default',
       'widow and happy', 'divorced and happy', 'single and happy',
       'single but default', 'married and default', 'widow but default',
       'divorced but default'], dtype=object)

__Default Rate Calculation__

In [67]:
data['marital_status_debt']
married_default_rate=len(data[data.marital_status_debt == 'married and default'])/(len(data[data.marital_status_debt == 'married and default'])+(len(data[data.marital_status_debt == 'married and happy'])))*100
married_default_rate

7.5451819434313965

In [68]:
data['marital_status_debt']
civil_default_rate=len(data[data.marital_status_debt == 'partner and default'])/(len(data[data.marital_status_debt == 'partner and default'])+(len(data[data.marital_status_debt == 'partner and happy'])))*100
civil_default_rate

9.347145266200915

In [69]:
data['marital_status_debt']
widow_er_default_rate=len(data[data.marital_status_debt == 'widow but default'])/(len(data[data.marital_status_debt == 'widow but default'])+(len(data[data.marital_status_debt == 'widow and happy'])))*100
widow_er_default_rate

6.569343065693431

In [70]:
data['marital_status_debt']
divorce_default_rate=len(data[data.marital_status_debt == 'divorced but default'])/(len(data[data.marital_status_debt == 'divorced but default'])+(len(data[data.marital_status_debt == 'divorced and happy'])))*100
divorce_default_rate

7.112970711297072

In [71]:
data['marital_status_debt']
single_default_rate=len(data[data.marital_status_debt == 'single but default'])/(len(data[data.marital_status_debt == 'single but default'])+(len(data[data.marital_status_debt == 'single and happy'])))*100
single_default_rate

9.750889679715302

In [72]:
marital_status_and_default={'Marital_Status':['married_default_rate','civil_default_rate','widow_er_default_rate','divorce_default_rate','single_default_rate'],
                           'Rate':[7.5,9.3,6.5,7.12,9.7]}
marital_status_and_default=pd.DataFrame(data=marital_status_and_default)
marital_status_and_default

Marital_Status  Rate
0   married_default_rate  7.50
1     civil_default_rate  9.30
2  widow_er_default_rate  6.50
3   divorce_default_rate  7.12
4    single_default_rate  9.70

__Answer:__ The data shows that the people who are single has more dendency to not pay debt on time followed by the civil partner.

#### Section 5.3: Is there a connection between income level and repaying a loan on time?<a class="anchor" id="section_5_3"></a>

In [73]:
income_debt=data[['income_level','debt']]
income_debt_relation=income_debt.groupby('income_level').mean().sort_values('debt',ascending=False)*100
income_debt_relation

debt
income_level              
middle_income     8.536436
lower_mid_income  8.536396
high_income       7.366109
low_income        6.256742

In [74]:
income_debt['debt'].value_counts()

0    19713
1     1741
Name: debt, dtype: int64

First we will observe the relation between income level and debt and make a new column with entries as below:
1. low pay no/have debt
2. good pay no/have debt
3. v.good pay no/have debt
4. high pay no/have debt

In [75]:
income_level_loan= [['income_level','debt']]


def income_level_rate(income_level_loan):
    level = income_level_loan['income_level']
    no_debts = income_level_loan['debt']
 
    if level == 'low_income':
        if no_debts == 0:
            return 'low pay no debt'
        return 'low pay have debt'
    if level == 'lower_mid_income':
        if no_debts == 0:
            return 'good pay no debt'
        return 'good pay have debt'
    if level == 'middle_income':
        if no_debts == 0:
            return 'v.good pay no debt'
        return 'v.good pay have debt'
    if level == 'high_income':
        if no_debts == 0:
            return 'High pay no debt'
        return 'High pay have debt'


data['income_level_debt']=data.apply(income_level_rate,axis=1)
data.head()

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   

                   purpose      income_level   category  \
0    purchase of the house       high_income   Property   
1        purchase of a car  lower_mid_income  Buy a Car   
2    purchase of the house     middle_income   Property   
3  supplementary education       high_income  Education   
4        to have a wedding     middle_income    Wedding   

   debt_children_category marital_status_debt   income_level_debt  
0  small family and happy   married and happy    High pay no debt  
1  small family and happy   married and happy    good pay no debt  
2  small family and happy   married and happy  v.good pay no debt  
3  large family and happy   married and happy    High pay no debt  
4  small family and happy   partner and happy  v.good pay no debt

__Default Rate Calculations:__
    

In [76]:

default_rate_high=len(data[data.income_level_debt == 'High pay have debt'])/(len(data[data.income_level_debt == 'High pay have debt'])+(len(data[data.income_level_debt == 'High pay no debt'])))*100

default_rate_good=len(data[data.income_level_debt == 'good pay have debt'])/(len(data[data.income_level_debt == 'good pay have debt'])+(len(data[data.income_level_debt == 'good pay no debt'])))*100

default_rate_v_good=len(data[data.income_level_debt == 'v.good pay have debt'])/(len(data[data.income_level_debt == 'v.good pay have debt'])+(len(data[data.income_level_debt == 'v.good pay no debt'])))*100

default_rate_low=len(data[data.income_level_debt == 'low pay have debt'])/(len(data[data.income_level_debt == 'low pay have debt'])+(len(data[data.income_level_debt == 'low pay no debt'])))*100


income_level_loan_repayment={'Income Level':['High Pay','V.Good Pay','Good Pay','Low Pay'],
                           'Default Rate':[default_rate_high,default_rate_v_good,default_rate_good,default_rate_low]}
income_and_default=pd.DataFrame(data=income_level_loan_repayment)
income_and_default


Income Level  Default Rate
0     High Pay      7.366109
1   V.Good Pay      8.536436
2     Good Pay      8.536396
3      Low Pay      6.256742

The middle income category has high default rate.

#### Section 5.4: How do different loan purposes affect on-time loan repayment?<a class="anchor" id="section_5_4"></a>

In [77]:
data['category'].unique()

array(['Property', 'Buy a Car', 'Education', 'Wedding'], dtype=object)

In [78]:
loan_type= [['category','debt']]


def loan_type_debt(loan_type):
    reason = loan_type['category']
    no_debts = loan_type['debt']
 
    if reason == 'Property':
        if no_debts == 0:
            return 'happily own property'
        return 'property and default'
    if reason == 'Education':
        if no_debts == 0:
            return 'happily educated'
        return 'educated and default'
    if reason == 'Wedding':
        if no_debts == 0:
            return 'happily married'
        return 'married and default'
    if reason == 'Buy a Car':
        if no_debts == 0:
            return 'happily own a car'
        return 'own a car and default'


data['loan_type_debt']=data.apply(loan_type_debt,axis=1)
data.head()

children  days_employed  dob_years            education  education_id  \
0         1           8437         42    bachelor's degree             0   
1         1           4024         36  secondary education             1   
2         0           5623         33  secondary education             1   
3         3           4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   

                   purpose      income_level   category  \
0    purchase of the house       high_income   Property   
1        purchase of a car  lower_mid_income  Buy a Car   
2    purchase of the house     middle_income   Property   
3  supplementary education       high_income  Education   
4        to have a wedding     middle_income    Wedding   

   debt_children_category marital_status_debt   income_level_debt  \
0  small family and happy   married and happy    High pay no debt   
1  small family and happy   married and happy    good pay no debt   
2  small family and happy   married and happy  v.good pay no debt   
3  large family and happy   married and happy    High pay no debt   
4  small family and happy   partner and happy  v.good pay no debt   

         loan_type_debt  
0  happily own property  
1     happily own a car  
2  happily own property  
3      happily educated  
4       happily married

In [79]:
data['loan_type_debt'].value_counts()
loan_real_default=len(data[data.loan_type_debt == 'property and default'])/(len(data[data.loan_type_debt == 'property and default'])+(len(data[data.loan_type_debt == 'happily own property'])))*100
loan_real_default
loan_education_default=len(data[data.loan_type_debt == 'educated and default'])/(len(data[data.loan_type_debt == 'educated and default'])+(len(data[data.loan_type_debt == 'happily educated'])))*100
loan_education_default
loan_car_default=len(data[data.loan_type_debt == 'own a car and default'])/(len(data[data.loan_type_debt == 'own a car and default'])+(len(data[data.loan_type_debt == 'happily own a car'])))*100
loan_car_default
loan_wedding_default=len(data[data.loan_type_debt == 'married and default'])/(len(data[data.loan_type_debt == 'married and default'])+(len(data[data.loan_type_debt == 'happily married'])))*100
loan_wedding_default

8.00344234079174

In [80]:
loan_type_default={'Laon Type':['Property','Education','Wedding','Car'],
                           'Rate':[7.2,9.2,8.0,9.3]}
Loan_type_default=pd.DataFrame(data=loan_type_default)
Loan_type_default

Laon Type  Rate
0   Property   7.2
1  Education   9.2
2    Wedding   8.0
3        Car   9.3

__Answer__ The default rate for property purposes loan is lowest and for car purposes is highest.

### Chapter 6: Conclusion <a class="anchor" id="chapter6"></a>

The aim of the project was to prepare a bank's loan division report. The data consisted of total 12 columns and 21525 no. of rows. After removing the duplicate values. The dataset consist of customers information regarding, children, marital status, level of education, income level and the purpose why they applied for the loan. We have categorize and observe different customers groups depending upon information given. Firstly, in preprocessed the data, missing values, data type,  duplicated values were handled and data was ready for the analysis. 

__Missing values handling:__

In missing values detection step, the column children of dataset there is typo and it seams that 1 and 20 were filled by -1 and 20 mistakenly. We will take absolute of the values to and replace with the actual corresponding value. Total 2174 values were missing in each column of days_employed and total_income. To deal with missing values the median were calculated and fill values afterward. In column of dob_years which represents the age of the customers,total 101 values missing and firstly the data was grouped in reference of the employment type and observed that customers are there are employed type or retiree. With median the age of all retiree and employed people were calculated which were 60 and 39 rexpectively. But some retiree people has age of 54 mentioned in the column so this approach was declined and overall median was calculated and fill values with that. The possible reason which was detected the this age group has mostly college or school level education. The reason for missing value whould be that relationship in very early age but this observation can be rejected on the basis when the customer failed to repay the loan. The relationship of age=0 values were observed with other columns. The mostly people has bachelor or secondary school education. One observation is that the these people are not really belong to one category but randomly distributed in different group of people.So, the possible reason could be that people were not willing to share date of birth or missid by the data entry person. The education columns values were string dtype str.lower method ws applied to remove the duplicate. The column named purchase has different reason for laon but the issue was detected with the description of same purpose in differnt word of choice. For example, car, purchase of car, Purchase of Car etc. represent the same purpose but word of description were different in many cases. A function was developed by categorizing the columns in car, house, eduation, real state and property. 

__Data Type__

The dataset dtype was changed from float to integer so that days_employed could be a even number rather than a fraction.


__Possible reason for duplicate values.__

There is a relation between duplicate values and the days_employed & total_income missing valaues. We observed this after filling valaues of days_employed & total_income that all the values we filled are now duplicate rows in dataset.

The reason would be when the data was collected the age was missing for some customers and again data has collected for these customers but the previous values were not updated or deleted after re enentring the missing values.

#### Data Categorization

__Categorizing the income_level column.__

The total_income was categorize in four levels.

1. Low income <=10,000
2. Low middle income <=20,000
3. Middle income <=30,000
4. High Income >30,000

__Categorizing the purpose of loan column.__

The general overview.

The first observation is that the purpose column look like a 'feedback/comment' option in form filling during loan process. There are different string values for the same purpose of laon, i.e. to have a wedding & having a wedding etc. There are many values i.e. car, car purpose, and cars.

The loan purpose dataset is now categorize in four main groups. All the entries related to wedding are put in wedding, laon for house, property and real state bases in property, for eduaction bases in education and all related to car were categorize in buy a car.


#### Credit Report

1. The analysis concluse that relation between default rate of both family groups is same.

2. The customers who are single has more dendency to not pay debt on time followed by the civil partner.

3. The middle income category has high default rate.

4. The default rate for property purposes loan is lowest and for car purposes is highest.
